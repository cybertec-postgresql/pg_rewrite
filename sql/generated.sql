-- Generated columns - some meaningful combinations of source and destination
-- columns.
CREATE TABLE tab7(
        i int primary key,
	j int,
	k int generated always as (i + 1) virtual,
	l int generated always AS (i + 1) stored,
	m int generated always AS (i + 1) virtual);
CREATE TABLE tab7_new(
        i int primary key,
        -- Override the value copied from the source table.
        j int generated always AS (i - 1) stored,
	-- Check that the expression is evaluated correctly on the source
	-- table.
	k int,
	-- The same for stored expression.
	l int,
	-- Override the value computed on the source table.
	m int generated always as (i - 1) virtual);
INSERT INTO tab7(i, j) VALUES (1, 1);
SELECT rewrite_table('tab7', 'tab7_new', 'tab7_orig');
SELECT * FROM tab7;

CREATE EXTENSION pageinspect;
-- HEAP_HASNULL indicates that the value of 'm' hasn't been copied from the
-- source table.
SELECT raw_flags
FROM heap_page_items(get_raw_page('tab7', 0)),
LATERAL heap_tuple_infomask_flags(t_infomask, t_infomask2);

-- For PG < 18, test without VIRTUAL columns.
CREATE TABLE tab8(
        i int primary key,
        j int,
        k int generated always AS (i + 1) stored);
CREATE TABLE tab8_new(
        i int primary key,
        -- Override the value copied from the source table.
        j int generated always AS (i - 1) stored,
        -- Check that the expression is evaluated correctly on the source
        -- table.
        k int);
INSERT INTO tab8(i, j) VALUES (1, 1);
SELECT rewrite_table('tab8', 'tab8_new', 'tab8_orig');
SELECT * FROM tab8;
