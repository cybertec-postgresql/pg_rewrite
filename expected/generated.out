-- Generated columns - some meaningful combinations of source and destination
-- columns.
CREATE TABLE tab7(
        i int primary key,
	j int,
	k int generated always as (i + 1) virtual,
	l int generated always AS (i + 1) stored,
	m int generated always AS (i + 1) virtual);
CREATE TABLE tab7_new(
        i int primary key,
        -- Override the value copied from the source table.
        j int generated always AS (i - 1) stored,
	-- Check that the expression is evaluated correctly on the source
	-- table.
	k int,
	-- The same for stored expression.
	l int,
	-- Override the value computed on the source table.
	m int generated always as (i - 1) virtual);
INSERT INTO tab7(i, j) VALUES (1, 1);
SELECT rewrite_table('tab7', 'tab7_new', 'tab7_orig');
 rewrite_table 
---------------
 
(1 row)

SELECT * FROM tab7;
 i | j | k | l | m 
---+---+---+---+---
 1 | 0 | 2 | 2 | 0
(1 row)

CREATE EXTENSION pageinspect;
-- HEAP_HASNULL indicates that the value of 'm' hasn't been copied from the
-- source table.
SELECT raw_flags
FROM heap_page_items(get_raw_page('tab7', 0)),
LATERAL heap_tuple_infomask_flags(t_infomask, t_infomask2);
                      raw_flags                       
------------------------------------------------------
 {HEAP_HASNULL,HEAP_XMIN_COMMITTED,HEAP_XMAX_INVALID}
(1 row)

-- For PG < 18, test without VIRTUAL columns.
CREATE TABLE tab8(
        i int primary key,
        j int,
        k int generated always AS (i + 1) stored);
CREATE TABLE tab8_new(
        i int primary key,
        -- Override the value copied from the source table.
        j int generated always AS (i - 1) stored,
        -- Check that the expression is evaluated correctly on the source
        -- table.
        k int);
INSERT INTO tab8(i, j) VALUES (1, 1);
SELECT rewrite_table('tab8', 'tab8_new', 'tab8_orig');
 rewrite_table 
---------------
 
(1 row)

SELECT * FROM tab8;
 i | j | k 
---+---+---
 1 | 0 | 2
(1 row)

